### ISSUE TITLE
Unordered-containers sometimes slower than hashmap

### ISSUE THREAD

**Issue by augustss** [tibbe/unordered-containers#119]
Here's a test program (instructions in the file), and input to the program (large).
The full log file was too big to upload to github in one go.
So unzip a?.zip, and then do 'cat xaa xab xac xad > qname.log'.

[Crunch.zip](https://github.com/tibbe/unordered-containers/files/117973/Crunch.zip)
[aa.zip](https://github.com/tibbe/unordered-containers/files/117979/aa.zip)
[ab.zip](https://github.com/tibbe/unordered-containers/files/117980/ab.zip)
[ac.zip](https://github.com/tibbe/unordered-containers/files/117978/ac.zip)
[ad.zip](https://github.com/tibbe/unordered-containers/files/117977/ad.zip)

**Comment by tibbe** [comment]
Thanks!

**Comment by augustss** [comment]
You won't thank me when you look at the test. :)

**Comment by augustss** [comment]
I forgot to mention that this is with ghc 7.8.4.

**Comment by augustss** [comment]
Here is a much simplified test.  Just a number of inserts and lookups.

[Test.zip](https://github.com/tibbe/unordered-containers/files/123242/Test.zip)

**Comment by augustss** [comment]
And an even simpler example!  It just has 1000 inserts.
It seem that when the hash of the key is large then hashmap is faster than unorderable-container.
By reducing the key values in this example you can see that happening.
The effect is even more pronounced when compiling with a 32 bit compiler.

[Test3.zip](https://github.com/tibbe/unordered-containers/files/123339/Test3.zip)

**Comment by augustss** [comment]
If I understand the code right, then data structure used here is not optimal if the hash values are spread over a large interval and sparse, since the depth of the tree will be log16 of the hash value.  It's an excellent data structure for dense arrays, though. 

The IntMap used by hashmap has a more flexible scheme and gets less depth.

**Comment by tibbe** [comment]
The data structures creates a new level if the lowest 4 bits in the hash collide (and the next 4 and so forth).

**Comment by augustss** [comment]
Yes, so hashes 0 and 2^31 together will create a "tree" 8 levels deep, right?  Which IntMap avoids by having a variable length prefix.

**Comment by tibbe** [comment]
Yes, that's right. We have the option to try to mix the hashes a bit (at a cost), to avoid some of the degenerate cases, but it's always a possibility.

**Comment by augustss** [comment]
As an experiment I did a mod with 100000 on the hash value.  With this change u-c is as fast as hashmap for the Mu compiler.  But reducing the hash values like that generates collisions, so it's really something that should be adapt automagically depending how full the hash table is.

**Comment by augustss** [comment]
Here's a picture of the biggest map from the benchmark.  A performance problem is the path from the root to where the fan-out starts.

![out](https://cloud.githubusercontent.com/assets/5245086/13004166/267e256c-d17a-11e5-8c1e-a985f3d1d298.png)

**Comment by augustss** [comment]
As an experiment I replaced the keys in the Crunch3 test with random keys, and then u-c beats hashmap with about 5%.

**Comment by tibbe** [comment]
We used to use a different hash function for `Int`s*. It was slower** and we didn't have good evidence that it was worth it, but perhaps we should try to go back to it and see if it helps in this case.

You can dig in the hashable repo for the Wang hash implementation and try that.

* Hash functions used in practice often go against what your typical CS text would tell you (i.e. make sure you have avalanche behavior). For example, Java, C++ (STL), and Python all use the identity function to hash ints. It's cheap and often the locality can actually be helpful. On the other hand hash table implementations in those languages sometimes do a bit extra mixing of hashes to avoid problems that stem from this practice (e.g. hashing pointers with the identity give you bad randomness in the lower bits).

** It doesn't help that GHC generates bad code whenever you try to implement simple hash functions in Haskell. For our previous int hash we had to resort to C to get good behavior for simple bit twiddling (!).

**Comment by tibbe** [comment]
By the way, that picture is very helpful. How did you generate it?

**Comment by augustss** [comment]
If there was some cheap way of randomizing the hash just before indexing it would help a lot (like N steps of a LFSR, but cheap).

Another thing to consider is to not use all bits in the hash for indexing.  You can start out with indexing with, say, 8 bits.  And then keep track of the number of collisions, and as the collisions go over some threshold you increase by 4 bits and rebuild the hash table (ugh!).

**Comment by augustss** [comment]
I wrote a mini version of u-c and hashmap to generate pictures.  I'll attach the code when I get home.

**Comment by augustss** [comment]
Here's the code to draw trees.


[draw.zip](https://github.com/tibbe/unordered-containers/files/129318/draw.zip)
